### PR TITLE
Blog app - integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by Mutai-Gilbert**
In this pull request we were able to:

- [x] Solved N+1 problem when fetching all posts and their comments for a user without using the bullet gem.

Screenshot before:
Screenshot 2023-06-20 at 21 45 09

Screenshot after:
Screenshot 2023-06-22 at 11 30 51

#### Integration specs

- User index page:
  - I can see the username of all other users.
  - I can see the profile picture for each user.
  - I can see the number of posts each user has written.
  - When I click on a user, I am redirected to that user's show page.

- user show page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see the user's bio.
  - I can see the user's first 3 posts.
  - I can see a button that lets me view all of a user's posts.
  - When I click a user's post, it redirects me to that post's show page.
  - When I click to see all posts, it redirects me to the user's post's index page.

- User post index page:
  - I can see the user's profile picture.
  - I can see the user's username.
  - I can see the number of posts the user has written.
  - I can see a post's title.
  - I can see some of the post's body.
  - I can see the first comments on a post.
  - I can see how many comments a post has.
  - I can see how many likes a post has.
  - I can see a section for pagination if there are more posts than fit on the view.
  - When I click on a post, it redirects me to that post's show page.

- Post show page:
  - I can see the post's title.
  - I can see who wrote the post.
  - I can see how many comments it has.
  - I can see how many likes it has.
  - I can see the post body.
  - I can see the username of each commentor.
  - I can see the comment each commentor left.

### CODE REVIEW

File: app/controllers/posts_controller.rb [hunk 1]
```diff
@@ -1,7 +1,6 @@
 class PostsController < ApplicationController
   def index
-    @user = User.find(params[:user_id])
-    @posts = Post.where(author_id: params[:user_id])
```
Comment (t-yanick) [inline app/controllers/posts_controller.rb:4]: - [x] Great job working on this milestone, trying to solve the N+1 problem and adding specs. I think the best approach in solving N+1 problem when fetching all posts and their comments for a user is by using `@posts = @user.posts.includes(:comments)` in the `posts#index`. Also in your index action in the posts_controller, you're not fetching posts but rather user that have made post and comments. Kindly review your approach and if possible take a new screenshot after trying the approach proposed.

File: spec/integration/post_show_spec.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+      @comment = Comment.create(author: @user, post: @first_post, text: 'This is my comment')
+      allow_any_instance_of(ApplicationController)
+        .to receive(:current_user)
+        .and_return(User.create(name: 'Tom', photo: 'https://unsplash.com/photos', bio: 'Teacher from Mexico.',
+                                post_counter: 0))
+      Like.create(author: @user, post: @first_post)
+      visit user_post_path(@user, @first_post)
+    end
+  end
+end
```
Comment (t-yanick) [inline spec/integration/post_show_spec.rb:40]: - [ ] Great job adding most of the tests. It seems you missed adding all the tests for the Post show page. Kindly add these tests.
image

**Review by t-yanick — CHANGES_REQUESTED** [review]
# STATUS: CHANGES REQUIRED

Hi @Mutai-Gilbert , 

### Good job so far!
### To Highlight! :clap:  :green_circle: 

:heavy_check_mark:  Effective use of capybara for testing
:heavy_check_mark: Solution screenshot of N+1 problem added to PR description
:heavy_check_mark: Use of correct Gitflow

There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- Making your feature branches name short and at the same time descriptive. I think `integration-specs-for-views-and-fix-n+1-problems` is too long

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/controllers/posts_controller.rb (hunk 1 above)
Comment (Mutai-Gilbert) [reply app/controllers/posts_controller.rb:4]: Done

**Review by ichala — APPROVED** [review]
Hi Student  ,
# Status: Approved :heavy_check_mark: :heavy_check_mark: :heavy_check_mark:  
 
68747470733a2f2f6d65646961322e67697068792e636f6d2f6d656469612f6c3366516333614356304871707a5763552f67697068792e676966

# To Highlight 💪🏻

- Changes Implemented ✔️ 


 Good job implementing the requirements  :+1: Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

# Optional Suggestions  :

- N/A


**********
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._
Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag  ichala in your question so I can receive the notification.


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**